iroh-net/src/magicsock.rs

```diff
@@ -348,18 +350,42 @@ impl Inner {
                             udp_sent = true;
                             // record metrics.
                         }
-                        Err(err) => {
+                        Poll::Ready(Err(err)) => {
                             error!(node = %public_key.fmt_short(), ?addr, "failed to send udp: {err:?}");
                             udp_error = Some(err);
                         }
+                        Poll::Pending => {
+                            udp_pending = true;
+                        }
                     }
                 }
 
                 // send derp
                 if let Some(ref derp_url) = derp_url {
-                    self.try_send_derp(derp_url, public_key, split_packets(&transmits));
-                    transmits_sent = transmits.len();
-                    derp_sent = true;
+                    match self.try_send_derp(derp_url, public_key, split_packets(&transmits)) {
+                        Poll::Ready(sent) => {
+                            derp_sent = sent;
+                            transmits_sent = transmits.len();
+                        }
+                        Poll::Pending => {
+                            self.network_send_wakers.lock().replace(cx.waker().clone());
+                            derp_pending = true;
+                        }
+                    }
+                }
+
+                if udp_addr.is_none() && derp_url.is_none() {
+                    // Handle no addresses being available
+                    warn!(node = %public_key.fmt_short(), "failed to send: no UDP or DERP addr");
+                    return Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::NotConnected,
+                        "no UDP or Derp address available for node",
+                    )));
+                }
+
+                if (udp_addr.is_none() || udp_pending) && (derp_url.is_none() || derp_pending) {
+                    // Handle backpressure
```

Comment:
Could we elaborate here on the tradeoff we're making?  IIUC we only return pending if we didn't manage to send to either.  That means we're accepting that some packets will only flow via one path and be dropped on the other.

However IIRC when we decide to send to both paths we do this because we're not certain that the UDP path is working.  So should we not always return pending if derp is pending in that case?

Reply:
Does this condition not also result in always returning pending when we have neither a derp nor a udp path?  However just below we're supposed to return a not-connected error in that case.

Reply:
> Does this condition not also result in always returning pending when we have neither a derp nor a udp path?

this is errored out in the above condition

Reply:
>  That means we're accepting that some packets will only flow via one path and be dropped on the other 

Yes, that's because we treat both connections as lossy, and sending is only best effort any way

Reply:
> > Does this condition not also result in always returning pending when we have neither a derp nor a udp path?
> 
> this is errored out in the above condition

oh yes, sorry i missed that.

Reply:
> > That means we're accepting that some packets will only flow via one path and be dropped on the other
> 
> Yes, that's because we treat both connections as lossy, and sending is only best effort any way

So your take is that we should not make sending via derp required when both paths are chosen?  I'm not 100% convinced that's the right choice.  In this situation either the udp path still works and recovers very quickly and will be marked the only path again.  Or it no longer works and will take longer and need holepunching etc again.  This could take up to 5s in the best case.  So I think I'm still in favour of requiring a derp send at this point, but if you really do not want to do this then I'll approve anyway.

But please update the comment making clear which way is chosen and why.  I think it would be good to have it described right there rather than having to find this conversation again.

Reply:
added a comment